cli/command/container/utils.go

```diff
@@ -13,28 +14,74 @@ import (
 	"golang.org/x/net/context"
 )
 
-func waitExitOrRemoved(ctx context.Context, dockerCli *command.DockerCli, containerID string, waitRemove bool) chan int {
+func waitExitOrRemoved(ctx context.Context, dockerCli *command.DockerCli, containerID string, attachDone chan struct{}, waitRemove bool) chan int {
 	if len(containerID) == 0 {
 		// containerID can never be empty
 		panic("Internal Error: waitExitOrRemoved needs a containerID as parameter")
 	}
 
-	var removeErr error
 	statusChan := make(chan int)
-	exitCode := 125
+	ctxWithCancel, cancel := context.WithCancel(ctx)
 
-	// Get events via Events API
 	f := filters.NewArgs()
 	f.Add("type", "container")
 	f.Add("container", containerID)
 	options := types.EventsOptions{
 		Filters: f,
 	}
+
+	go func() {
+		statusChan <- waitExitedOrRemovedEvent(ctxWithCancel, dockerCli, containerID, waitRemove, options)
+		cancel()
+	}()
+
+	if waitRemove {
+		// This is a fallback in case there are issues with the events API.
+		// Uses polling at pretty long intervals instead of events
+		// This makes sure we don't hang the CLI.
+		go func() {
+			select {
+			case <-attachDone:
+			case <-ctxWithCancel.Done():
+				return
+			}
+
+			waitRemovedPolling(ctxWithCancel, dockerCli, containerID)
+			cancel()
+
+			// try to get a replay of events so we can collect the exit status
+			// but still timeout in case of missed events
+			ctxWithTimeout, cancel := context.WithTimeout(ctx, 30*time.Second)
+			options.Since = "0"
+			statusChan <- waitExitedOrRemovedEvent(ctxWithTimeout, dockerCli, containerID, false, options)
+			cancel()
+		}()
+	}
+
+	return statusChan
+}
+
+func waitRemovedPolling(ctx context.Context, dockerCli *command.DockerCli, containerID string) {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			_, err := dockerCli.Client().ContainerInspect(ctx, containerID)
+			if err != nil && clientapi.IsErrNotFound(err) {
+				return
+			}
+		}
+	}
+}
+
+func waitExitedOrRemovedEvent(ctx context.Context, dockerCli *command.DockerCli, containerID string, waitRemove bool, eventOpts types.EventsOptions) (exitCode int) {
 	eventCtx, cancel := context.WithCancel(ctx)
-	eventq, errq := dockerCli.Client().Events(eventCtx, options)
+	eventq, errq := dockerCli.Client().Events(eventCtx, eventOpts)
 
+	exitCode = 125 // set default exit code
 	eventProcessor := func(e events.Message) bool {
```

Comment:
process a single event and indicate if we've received an event telling us to finish or not.